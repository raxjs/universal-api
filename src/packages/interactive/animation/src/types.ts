/// <reference path='../../../../../types/interface.d.ts'/>

type TimingFunction = 'linear' | 'ease' | 'ease-in' | 'ease-in-out' | 'ease-out' | 'step-start' | 'step-end' | string;

export interface AnimationOptions extends Uni.COptions {
  duration?: number;

  timingFunction?: TimingFunction;

  delay?: number;

  transformOrigin?: string;
}

export interface AnimationActionAnimate {
  type: 'style'
  | 'rotate' | 'rotate3d' | 'rotateX' | 'rotateY' | 'rotateZ'
  | 'scale' | 'scale3d' | 'scaleX' | 'scaleY' | 'scaleZ'
  | 'translate' | 'translate3d' | 'translateX' | 'translateY' | 'translateZ'
  | 'skew' | 'skewX' | 'skewY'
  | 'matrix' | 'matrix3d';
  args: any[];
}

export interface AnimationAction {
  animates: AnimationActionAnimate[];
  option: {
    transformOrigin: string;
    transition: {
      delay: number;
      duration: number;
      timingFunction: TimingFunction;
    };
  };
}

export interface AnimationData {
  actions: AnimationAction[];
}

export interface Animation {

  /* ******************* util ******************* */

  /**
   * Export the animation queue, the previous animation data will be cleared after each call of the `export` method.
   * If you need to support the web platform, the parameter `dom` needs to be passed in, otherwise you donâ€™t need to pass in.
   * @param dom
   */
  export: (dom?: HTMLElement) => AnimationData;

  /**
   * Indicates that a set of animations are completed.
   * Any number of animation methods can be called in a group of animations, all animations in a group of animations will start at the same time,
   * and the next group of animations will be executed after a group of animations are completed.
   * @param options Same as the parameters of the `createAnimation` method
   */
  step: (options?: AnimationOptions) => Animation;


  /* ******************* style ******************* */

  /**
   * Set opacity
   * @param value
   */
  opacity: (value: number) => Animation;

  /**
   * Set background color
   * @param value
   */
  backgroundColor: (value: string) => Animation;

  /**
   * Set width
   * @param value
   */
  width: (value: number | string) => Animation;

  /**
   * Set height
   * @param value
   */
  height: (value: number | string) => Animation;

  /**
   * Set the `left` value
   * @param value
   */
  left: (value: number | string) => Animation;

  /**
   * Set the `right` value
   * @param value
   */
  right: (value: number | string) => Animation;

  /**
   * Set the `top` value
   * @param value
   */
  top: (value: number | string) => Animation;

  /**
   * Set the `bottom` value
   * @param value
   */
  bottom: (value: number | string) => Animation;


  /* ******************* rotate ******************* */

  /**
   * Rotate an angle clockwise from the origin
   * @param angle
   */
  rotate: (angle: number) => Animation;

  /**
   * Rotate one angle clockwise from the X axis
   * @param angle
   */
  rotateX: (angle: number) => Animation;

  /**
   * Rotate one angle clockwise from the Y axis
   * @param angle
   */
  rotateY: (angle: number) => Animation;

  /**
   * Rotate one angle clockwise from the Z axis
   * @param angle
   */
  rotateZ: (angle: number) => Animation;

  /**
   * Rotate one angle clockwise from the fixed axis
   * @param x
   * @param y
   * @param z
   * @param angle
   */
  rotate3d: (x: number, y: number, z: number, angle: number) => Animation;


  /* ******************* scale ******************* */

  /**
   * Scale X and Y axes
   * @param sx
   * @param sy
   */
  scale: (sx: number, sy: number) => Animation;

  /**
   * Scale X axes
   * @param scale
   */
  scaleX: (scale: number) => Animation;

  /**
   * Scale Y axes
   * @param scale
   */
  scaleY: (scale: number) => Animation;

  /**
   * Scale Z axes
   * @param scale
   */
  scaleZ: (scale: number) => Animation;

  /**
   * Scale X, Y and Z axis
   * @param sx
   * @param sy
   * @param sz
   */
  scale3d: (sx: number, sy: number, sz: number) => Animation;


  /* ******************* translate ******************* */

  /**
   * Translate transformation
   * @param tx
   * @param ty
   */
  translate: (tx: number, ty: number) => Animation;

  /**
   * Translate the X axis
   * @param translation
   */
  translateX: (translation: number) => Animation;

  /**
   * Translate the Y axis
   * @param translation
   */
  translateY: (translation: number) => Animation;

  /**
   * Translate the Z axis
   * @param translation
   */
  translateZ: (translation: number) => Animation;

  /**
   * Translate the X, Y and Z axis
   * @param tx
   * @param ty
   * @param tz
   */
  translate3d: (tx: number, ty: number, tz: number) => Animation;


  /* ******************* skew ******************* */

  /**
   * Tilt X and Y axis
   * @param ax
   * @param ay
   */
  skew: (ax: number, ay: number) => Animation;

  /**
   * Tilt X axis
   * @param angle
   */
  skewX: (angle: number) => Animation;

  /**
   * Tilt Y axis
   * @param angle
   */
  skewY: (angle: number) => Animation;


  /* ******************* matrix ******************* */

  /**
   * Matrix transformation
   * @param a
   * @param b
   * @param c
   * @param d
   * @param tx
   * @param ty
   * @see https://developer.mozilla.org/en-US/docs/Web/CSS/transform-function/matrix()
   */
  matrix: (
    a: number, b: number, c: number, d: number,
    tx: number, ty: number
  ) => Animation;

  /**
   * 3D matrix transformation
   * @param a1
   * @param b1
   * @param c1
   * @param d1
   * @param a2
   * @param b2
   * @param c2
   * @param d2
   * @param a3
   * @param b3
   * @param c3
   * @param d3
   * @param a4
   * @param b4
   * @param c4
   * @param d4
   * @see https://developer.mozilla.org/en-US/docs/Web/CSS/transform-function/matrix3d()
   */
  matrix3d: (
    a1: number, b1: number, c1: number, d1: number,
    a2: number, b2: number, c2: number, d2: number,
    a3: number, b3: number, c3: number, d3: number,
    a4: number, b4: number, c4: number, d4: number
  ) => Animation;
}

interface TransitionStyleOptions {
  opacity?: number;
  backgroundColor?: string;
  width?: number | string;
  height?: number | string;
  left?: number | string;
  right?: number | string;
  top?: number | string;
  bottom?: number | string;
  transform?: string;
}

export interface TransitionOptions {
  from?: TransitionStyleOptions;
  to?: TransitionStyleOptions;
  options?: AnimationOptions;
}
