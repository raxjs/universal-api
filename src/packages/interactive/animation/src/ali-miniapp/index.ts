import createAnimation from './createAnimation';

export {
  createAnimation,
};

export default {
  createAnimation,
};
