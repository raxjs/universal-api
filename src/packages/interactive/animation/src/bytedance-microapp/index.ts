import { Animation, AnimationOptions } from '../types';

export function createAnimation(options?: AnimationOptions): Animation {
  return tt.createAnimation(options);
}

export default {
  createAnimation,
};
