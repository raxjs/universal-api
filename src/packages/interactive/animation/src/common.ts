import { styleIn } from '@utils/styleOptions';
import { AnimationOptions } from './types';

/**
 * merge default options
 * @param containerName
 * @param options
 */
export function getDefaultOptions(containerName: string, options?: AnimationOptions): AnimationOptions {
  return styleIn(
    {
      duration: 400,
      timingFunction: 'linear',
      delay: 0,
      transformOrigin: '50% 50% 0',
      ...options,
    },
    containerName,
  );
}
