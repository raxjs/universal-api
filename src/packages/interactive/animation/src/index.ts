import { isMiniApp, isWeChatMiniProgram, isWeb, isByteDanceMicroApp } from '@uni/env';
import aliMiniAppModule from './ali-miniapp/index';
import webModule from './web/index';
import weChatModule from './wechat-miniprogram/index';
import bytedanceModule from './bytedance-microapp/index';
import { AnimationOptions, TransitionOptions } from './types';

export function createAnimation(options?: AnimationOptions) {
  if (isWeb) {
    return webModule.createAnimation(options);
  } else if (isWeChatMiniProgram) {
    return weChatModule.createAnimation(options);
  } else if (isByteDanceMicroApp) {
    return bytedanceModule.createAnimation(options);
  } else if (isMiniApp) {
    return aliMiniAppModule.createAnimation(options);
  } else {
    throw new Error('@uni/apis：createAnimation 暂不支持');
  }
}

export function createTransition(options: TransitionOptions) {
  if (isWeb) {
    return webModule.createTransition(options);
  } else if (isWeChatMiniProgram) {
    return weChatModule.createTransition(options);
  } else if (isByteDanceMicroApp) {
    return bytedanceModule.createTransition(options);
  } else if (isMiniApp) {
    return aliMiniAppModule.createTransition(options);
  } else {
    throw new Error('@uni/apis：createTransition 暂不支持');
  }
}

export default {
  createAnimation,
  createTransition,
};
