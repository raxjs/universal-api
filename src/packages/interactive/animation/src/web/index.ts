import { Animation, AnimationOptions } from '../types';
import AnimationImpl from './AnimationImpl';

export function createAnimation(options?: AnimationOptions): Animation {
  return new AnimationImpl(options);
}

export default {
  createAnimation,
};
