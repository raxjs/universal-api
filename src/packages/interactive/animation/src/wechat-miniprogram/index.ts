import { Animation, AnimationOptions } from '../types';

export function createAnimation(options?: AnimationOptions): Animation {
  return wx.createAnimation(options);
}

export default {
  createAnimation,
};
