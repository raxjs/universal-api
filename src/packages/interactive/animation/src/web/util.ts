import { AnimationAction, AnimationOptions } from '../types';
import { normalizeOptions } from '../common';
import { CONTAINER_NAME } from '@utils/constant';

/**
 * normalize unit
 * @param value
 * @param unit
 * @param force
 */
export function normalizeUnit(value: number | string, unit: string, force = false): string {
  if (force || typeof value === 'number') {
    return `${value}${unit}`;
  }
  return value;
}

/**
 * merge default options
 * @param options
 */
export function getDefaultOptions(options?: AnimationOptions): AnimationOptions {
  return normalizeOptions(
    {
      duration: 400,
      timingFunction: 'linear',
      delay: 0,
      transformOrigin: '50% 50% 0',
      ...options,
    },
    CONTAINER_NAME.WEB,
  );
}

/**
 * handle animation actions queue
 * @param actions
 * @param fn
 * @param onFinish
 */
export function handleActionsQueue(
  actions: AnimationAction[],
  fn: (action: AnimationAction, callback: () => void) => any,
  onFinish?: () => any,
) {
  actions = actions.slice();
  const action = actions.shift();
  if (action) {
    fn(action, () => handleActionsQueue(actions, fn, onFinish));
  } else if (onFinish) {
    onFinish();
  }
}
