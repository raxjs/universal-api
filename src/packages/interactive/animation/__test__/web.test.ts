import { sleep, testWebAPI } from '@/utils/__test__/util';
import { Animation } from '../src/types';

testWebAPI('animation', async () => {
  /* eslint-disable newline-per-chained-call */

  const { createAnimation, createTransition } = require('../src/index.ts');

  let dom = document.createElement('div');
  const animation: Animation = createAnimation({ duration: 50 });
  animation.top(1).opacity(0.5).backgroundColor('#aaa').width(100).height(200).left(300).right(400).top(500).bottom(600).step();
  animation.rotate(1).rotateX(2).rotateY(3).rotateZ(4).rotate3d(1, 2, 3, 4).step();
  animation.scale(1, 2).scaleX(2).scaleY(3).scaleZ(4).scale3d(1, 2, 3).step();
  animation.translate(1, 2).translateX(2).translateY(3).translateZ(4).translate3d(1, 2, 3).step();
  animation.skew(1, 2).skewX(2).skewY(3).step();
  animation.matrix(1, 2, 3, 4, 5, 6).matrix3d(1, 2, 3, 4, 1, 2, 3, 4, 1, 2, 3, 4, 1, 2, 3, 4).step();

  animation.export(dom);
  // for six step animation, per step duration is 100ms
  await sleep(4);
  expect(dom).toMatchSnapshot();
  await sleep(50);
  expect(dom).toMatchSnapshot();
  await sleep(50);
  expect(dom).toMatchSnapshot();
  await sleep(50);
  expect(dom).toMatchSnapshot();
  await sleep(50);
  expect(dom).toMatchSnapshot();
  await sleep(50);
  expect(dom).toMatchSnapshot();

  // update the animation while the animation is running
  dom = document.createElement('div');
  dom.style.transitionTimingFunction = 'ease';

  createAnimation({ duration: 50 }).scale(2, 4).top(100).step().export(dom);
  await sleep(4);
  expect(dom).toMatchSnapshot();

  createAnimation({ duration: 50 }).skew(3, 6).bottom(200).step().export(dom);
  await sleep(4);
  expect(dom).toMatchSnapshot();

  // for restore style
  await sleep(50);
  expect(dom).toMatchSnapshot();


  // createTransition
  dom = document.createElement('div');
  createTransition({
    from: {
      transform: 'rotate(10deg) translateX(10px)',
      opacity: 0.5,
    },
    to: {
      transform: 'rotate(20deg) translateX(20px)',
      opacity: 1,
    },
    options: {
      duration: 50,
    },
  }).export(dom);
  await sleep(4);
  expect(dom).toMatchSnapshot();
  await sleep(16);
  expect(dom).toMatchSnapshot();

  /* eslint-enable newline-per-chained-call */
});
