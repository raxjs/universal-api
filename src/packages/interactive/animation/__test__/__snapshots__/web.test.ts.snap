// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Platform API: animation Test container: web 1`] = `
<div
  style="top: 500px; opacity: 0.5; background-color: rgb(170, 170, 170); width: 100px; height: 200px; left: 300px; right: 400px; bottom: 600px; transition-property: all; transition-duration: 50ms; transition-delay: 0ms; transition-timing-function: linear; transform-origin: 50% 50% 0;"
/>
`;

exports[`Test Platform API: animation Test container: web 2`] = `
<div
  style="top: 500px; opacity: 0.5; background-color: rgb(170, 170, 170); width: 100px; height: 200px; left: 300px; right: 400px; bottom: 600px; transition-property: all; transition-duration: 50ms; transition-delay: 0ms; transition-timing-function: linear; transform-origin: 50% 50% 0; transform: rotate(1deg) rotateX(2deg) rotateY(3deg) rotateZ(4deg) rotate3d(1,2,3,4deg);"
/>
`;

exports[`Test Platform API: animation Test container: web 3`] = `
<div
  style="top: 500px; opacity: 0.5; background-color: rgb(170, 170, 170); width: 100px; height: 200px; left: 300px; right: 400px; bottom: 600px; transition-property: all; transition-duration: 50ms; transition-delay: 0ms; transition-timing-function: linear; transform-origin: 50% 50% 0; transform: rotate(1deg) rotateX(2deg) rotateY(3deg) rotateZ(4deg) rotate3d(1,2,3,4deg) scale(1,2) scaleX(2) scaleY(3) scaleZ(4) scale3d(1,2,3);"
/>
`;

exports[`Test Platform API: animation Test container: web 4`] = `
<div
  style="top: 500px; opacity: 0.5; background-color: rgb(170, 170, 170); width: 100px; height: 200px; left: 300px; right: 400px; bottom: 600px; transition-property: all; transition-duration: 50ms; transition-delay: 0ms; transition-timing-function: linear; transform-origin: 50% 50% 0; transform: rotate(1deg) rotateX(2deg) rotateY(3deg) rotateZ(4deg) rotate3d(1,2,3,4deg) scale(1,2) scaleX(2) scaleY(3) scaleZ(4) scale3d(1,2,3) translate(1px,2px) translateX(2px) translateY(3px) translateZ(4px) translate3d(1px,2px,3px);"
/>
`;

exports[`Test Platform API: animation Test container: web 5`] = `
<div
  style="top: 500px; opacity: 0.5; background-color: rgb(170, 170, 170); width: 100px; height: 200px; left: 300px; right: 400px; bottom: 600px; transition-property: all; transition-duration: 50ms; transition-delay: 0ms; transition-timing-function: linear; transform-origin: 50% 50% 0; transform: rotate(1deg) rotateX(2deg) rotateY(3deg) rotateZ(4deg) rotate3d(1,2,3,4deg) scale(1,2) scaleX(2) scaleY(3) scaleZ(4) scale3d(1,2,3) translate(1px,2px) translateX(2px) translateY(3px) translateZ(4px) translate3d(1px,2px,3px) skew(1deg,2deg) skewX(2deg) skewY(3deg);"
/>
`;

exports[`Test Platform API: animation Test container: web 6`] = `
<div
  style="top: 500px; opacity: 0.5; background-color: rgb(170, 170, 170); width: 100px; height: 200px; left: 300px; right: 400px; bottom: 600px; transition-property: all; transition-duration: 50ms; transition-delay: 0ms; transition-timing-function: linear; transform-origin: 50% 50% 0; transform: rotate(1deg) rotateX(2deg) rotateY(3deg) rotateZ(4deg) rotate3d(1,2,3,4deg) scale(1,2) scaleX(2) scaleY(3) scaleZ(4) scale3d(1,2,3) translate(1px,2px) translateX(2px) translateY(3px) translateZ(4px) translate3d(1px,2px,3px) skew(1deg,2deg) skewX(2deg) skewY(3deg) matrix(1,2,3,4,5,6) matrix3d(1,2,3,4,1,2,3,4,1,2,3,4,1,2,3,4);"
/>
`;

exports[`Test Platform API: animation Test container: web 7`] = `
<div
  style="transition-timing-function: linear; top: 100px; transform: scale(2,4); transition-property: all; transition-duration: 50ms; transition-delay: 0ms; transform-origin: 50% 50% 0;"
/>
`;

exports[`Test Platform API: animation Test container: web 8`] = `
<div
  style="transition-timing-function: linear; top: 100px; transform: skew(3deg,6deg); transition-property: all; transition-duration: 50ms; transition-delay: 0ms; transform-origin: 50% 50% 0; bottom: 200px;"
/>
`;

exports[`Test Platform API: animation Test container: web 9`] = `
<div
  style="transition-timing-function: ease; top: 100px; transform: skew(3deg,6deg); bottom: 200px;"
/>
`;

exports[`Test Platform API: animation Test container: web 10`] = `
<div
  style="opacity: 0.5; transform: rotate(10deg) translateX(10px); transition-property: all; transition-duration: 16ms; transition-delay: 0ms; transition-timing-function: linear; transform-origin: 50% 50% 0;"
/>
`;

exports[`Test Platform API: animation Test container: web 11`] = `
<div
  style="opacity: 1; transform: rotate(20deg) translateX(20px); transition-property: all; transition-duration: 50ms; transition-delay: 0ms; transition-timing-function: linear; transform-origin: 50% 50% 0;"
/>
`;
